modelci/finetuner/trainer.py

```diff
@@ -6,253 +6,85 @@
-            self,
-            net: nn.Module,
-            optimizer: torch.optim.optimizer.Optimizer,
-            train_data_loader: torch.utils.data.DataLoader,
-            test_data_loader: torch.utils.data.DataLoader,
-            train_data_size: int = None,
-            test_data_size: int = None,
-            scheduler: torch.optim.lr_scheduler._LRScheduler = None,  # noqa
-            device: torch.device = None,
-    ):
-        """
-        Args:
-            net (nn.Module): PyTorch model.
-            train_data_loader (torch.utils.data.DataLoader): Data loader for training.
-            test_data_loader (torch.utils.data.DataLoader): Data loader for evaluating.
-            optimizer (torch.optim.optimizer.Optimizer): Optimizer.
-            train_data_size (int): Size of training data. Optional. Needed when using a sampler to split
-                training and validation data.
-            test_data_size (int): Size of evaluation data. Optional. Needed when using a sampler to split
-                training and validation data.
-        """
-        super().__init__(
-            net, optimizer=optimizer, train_data_loader=train_data_loader, test_data_loader=test_data_loader,
-            train_data_size=train_data_size, test_data_size=test_data_size, device=device,
-        )
-        self.scheduler = scheduler
-
-    def after_batch_finish(self):
-        if self.scheduler:
-            self.scheduler.step()
+    def __init__(self, model: pl.LightningModule, data_loader_kwargs: dict = None, trainer_kwargs: dict = None):
+        self.model = model
+        trainer_kwargs = trainer_kwargs or dict()
+        self.trainer_engine = pl.Trainer(**trainer_kwargs)
+        self._data_loader_kwargs = data_loader_kwargs or dict()
 
-    def train_one_batch(self, samples):
-        """Train one epoch.
+        self._executor = ThreadPoolExecutor(max_workers=1)
+        self._event_pause = threading.Event()
+        self._task: Optional[Future] = None
 
-        Args:
-            samples: One batch of data
+    def start(self):
+        self._task = self._executor.submit(self.trainer_engine.fit, self.model, **self._data_loader_kwargs)
 
-        Returns:
-            Tuple of training loss and number of correctly predicted examples
-        """
-        inputs, labels = samples
-        inputs = inputs.to(self.device)
-        labels = labels.to(self.device)
+    def join(self, timeout=None):
+        if self._task:
+            self._task.result(timeout=timeout)
 
-        # forward + backward + optimize
-        outputs = self.net(inputs)
-        loss = self.criterion(outputs, labels)
-        loss.backward()
-        self.optimizer.step()
-        predicted_labels = torch.argmax(outputs, dim=1)
-        corrects = torch.sum(predicted_labels == labels)
+    def terminate(self):
+        if self._task:
+            # trigger pytorch lighting training graceful shutdown via a ^C
+            self._task.set_exception(KeyboardInterrupt())
```

Comment:
Does users manually do this?

Reply:
Yes, this is to kill/delete a training task